private/buf/bufsync/bufsyncapi/sync_handler.go

```diff
@@ -0,0 +1,526 @@
+			)),
+		)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *syncHandler) IsProtectedBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	// If the branch is the Git default branch, protect it.
+	if branchName == h.repo.DefaultBranch() {
+		return true, nil
+	}
+	return h.IsReleaseBranch(ctx, moduleIdentity, branchName)
+}
+
+func (h *syncHandler) IsReleaseBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	// Otherwise the only other protected branch is the Repository's default (release) branch.
+	// We cache a repository's default branch even though it can change because it's _extremely_ unlikely that it changes.
+	cacheKey := moduleIdentity.IdentityString()
+	if _, ok := h.moduleIdentityToDefaultBranchCache[cacheKey]; !ok {
+		service := h.repositoryServiceClientFactory(moduleIdentity.Remote())
+		res, err := service.GetRepositoryByFullName(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryByFullNameRequest{
+			FullName: moduleIdentity.Owner() + "/" + moduleIdentity.Repository(),
+		}))
+		if err != nil {
+			if connect.CodeOf(err) == connect.CodeNotFound {
+				// Repo not created, no branch is protected because no branches exist. We cache this
+				// because it shouldn't change during the lifetime of sync.
+				h.moduleIdentityToDefaultBranchCache[cacheKey] = ""
+			}
+			return false, fmt.Errorf("load repository %q: %w", cacheKey, err)
+		}
+		h.moduleIdentityToDefaultBranchCache[cacheKey] = res.Msg.Repository.DefaultBranch
+	}
+	return branchName == h.moduleIdentityToDefaultBranchCache[cacheKey], nil
+}
+
+func (h *syncHandler) GetBranchHead(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (*registryv1alpha1.RepositoryCommit, error) {
+	repositoryID, err := h.getRepositoryID(ctx, moduleIdentity)
+	if err != nil {
+		return nil, err
+	}
+	service := h.repositoryBranchServiceClientFactory(moduleIdentity.Remote())
+	branchRes, err := service.GetRepositoryBranch(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryBranchRequest{
+		RepositoryId: repositoryID,
+		Name:         branchName,
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			return nil, nil
+		}
+		return nil, err
+	}
+	commitName := branchRes.Msg.Branch.LatestCommitName
+	if commitName == "" {
+		return nil, nil // branch has no commits on it
+	}
+	commitService := h.repositoryCommitServiceClientFactory(moduleIdentity.Remote())
+	res, err := commitService.GetRepositoryCommitByReference(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryCommitByReferenceRequest{
+		RepositoryOwner: moduleIdentity.Owner(),
+		RepositoryName:  moduleIdentity.Repository(),
+		Reference:       commitName,
+	}))
+	if err != nil {
+		return nil, err
+	}
+	return res.Msg.RepositoryCommit, nil
```

Comment:
Missing to check that the returned reference was a BSR commit?

Reply:
What do you mean? What else could it be?

Reply:
It could be a tag with the same name as a commit, although our resolve logic checks commits first, so nevermind.